morph_tool/morphdb.py

```diff
@@ -0,0 +1,346 @@
+    df.astype({key: bool for key in BOOLEAN_REPAIR_ATTRS}, copy=False)
+    return df
+
+
+class MorphDB(object):
+    '''A MorphInfo container.
+    It takes care of maintaining unicity of the MorphInfo element
+    and methods to write neurondb to various format (xml, dat, csv)
+    '''
+
+    def __init__(self):
+        self.df = pd.DataFrame(columns=COLUMNS)
+        self.df.astype({key: bool for key in BOOLEAN_REPAIR_ATTRS}, copy=False)
+
+    @classmethod
+    def from_neurondb(cls,
+                      neurondb: Path = None,
+                      label: str = 'default',
+                      morphology_folder: Optional[Path] = None):
+        '''Builds a MorphologyDB from a neurondb.(xml|dat) file
+        Args:
+            neurondb: path to a neurondb.(xml|dat) file
+            label: a unique label to mark all morphologies coming from this neurondb
+            morphology_folder: the location of the morphology files, if None it will default
+                to the neurondb folder
+
+        ..note:: missing keys are filled with `True` values
+        '''
+        obj = MorphDB()
+
+        if not morphology_folder:
+            morphology_folder = neurondb.parent.resolve()
+
+        if neurondb.suffix.lower() == '.dat':
+            columns = ['name', 'layer', 'mtype']
+            obj.df = pd.read_csv(neurondb, sep=r'\s+', names=columns, usecols=range(len(columns)))
+
+            fulltypes = obj.df.mtype.str.split(':', n=1, expand=True).fillna('')
+            if len(fulltypes.columns) > 1:
+                obj.df['mtype_no_subtype'] = fulltypes[0]
+                obj.df['msubtype'] = fulltypes[1]
+            else:
+                obj.df['mtype_no_subtype'] = obj.df.mtype
+                obj.df['msubtype'] = ''
+            obj.df['label'] = label
+            for missing_col in set(COLUMNS) - set(obj.df.columns):
+                obj.df[missing_col] = None
+            obj.df.layer = obj.df.layer.astype('str')
+            obj.df['path'] = obj.df.apply(lambda row: find_morph(
+                morphology_folder, row['name']), axis=1)
+            obj.df = obj.df.reindex(columns=COLUMNS)
+            for key in BOOLEAN_REPAIR_ATTRS:
+                obj.df[key] = True
+            obj.df['axon_inputs'] = [[] for _ in range(len(obj.df))]
+
+            return obj
+
+        with neurondb.open() as fd:
+            content = fd.read()
+            neurondb = xmltodict.parse(content)
+
+        morphologies = neurondb['neurondb']['listing']['morphology']
+
+        # Case where there is a single <morphology></morphology> tag in the XML
+        if not isinstance(morphologies, list):
+            morphologies = [morphologies]
+        morphologies = filter(None, morphologies)
+        morphologies = list(map(MorphInfo._from_xml, morphologies))  # noqa, pylint: disable=protected-access
+
+        for morph in morphologies:
+            morph.label = label
+            morph.path = find_morph(morphology_folder, morph.name)
```

Comment:
I also added a line to put the abstolute morphology path as another column, would it make sense to have that? So i'll be easier later to load morphs, withoutt having to remember where the base_path is?

Reply:
This is a `Path` object so you should not need the absolute path I think.

Reply:
sorry, I was thinking of saving data to .csv, which I can carry around, so I need the absolute path saved in it.

Reply:
Ah, yes, got it. Indeed, good point !

Reply:
I think with the current way, they already are absolute paths.

Reply:
yes indeed, I just checked to be sure, all good!